src/acl/Asn.cc

```diff
@@ -318,57 +313,35 @@ asHandleReply(void *data, StoreIOBuffer result)
 
         *t = '\0';
         debugs(53, 3, "asHandleReply: AS# " << s << " (" << asState->as_number << ")");
-        asnAddNet(s, asState->as_number);
+        if (asState->leftovers.contentSize()) {
+            asState->leftovers.append(s, strlen(s));
+            asState->leftovers.terminate();
+            asnAddNet(asState->leftovers.content(), asState->as_number);
+            asState->leftovers.reset();
+        } else {
+            asnAddNet(s, asState->as_number);
+        }
         s = t + 1;
-        asState->dataRead = true;
     }
 
     /*
      * Next, grab the end of the 'valid data' in the buffer, and figure
      * out how much data is left in our buffer, which we need to keep
      * around for the next request
      */
-    leftoversz = (asState->reqofs + result.length) - (s - buf);
+    leftoversz = result.length - (s - buf);
 
     assert(leftoversz >= 0);
 
-    /*
-     * Next, copy the left over data, from s to s + leftoversz to the
-     * beginning of the buffer
-     */
-    memmove(buf, s, leftoversz);
-
-    /*
-     * Next, update our offset and reqofs, and kick off a copy if required
-     */
-    asState->offset += result.length;
-
-    asState->reqofs = leftoversz;
-
-    debugs(53, 3, "asState->offset = " << asState->offset);
-
-    if (e->store_status == STORE_PENDING) {
-        debugs(53, 3, "asHandleReply: store_status == STORE_PENDING: " << e->url()  );
-        StoreIOBuffer tempBuffer (AS_REQBUF_SZ - asState->reqofs,
-                                  asState->offset,
-                                  asState->reqbuf + asState->reqofs);
-        storeClientCopy(asState->sc,
-                        e,
-                        tempBuffer,
-                        asHandleReply,
-                        asState);
-    } else {
-        StoreIOBuffer tempBuffer;
-        debugs(53, 3, "asHandleReply: store complete, but data received " << e->url()  );
-        tempBuffer.offset = asState->offset;
-        tempBuffer.length = AS_REQBUF_SZ - asState->reqofs;
-        tempBuffer.data = asState->reqbuf + asState->reqofs;
-        storeClientCopy(asState->sc,
-                        e,
-                        tempBuffer,
-                        asHandleReply,
-                        asState);
+    if (leftoversz > 0) {
+        asState->leftovers.reset(); // must be empty already
+        asState->leftovers.append(s, leftoversz); // save the left over data, from s to s + leftoversz
     }
+
+    debugs(53, 3, (e->store_status == STORE_PENDING ? "STORE_PENDING" : "STORE_OK") << " " << e->url());
+
+    if (!result.flags.eof)
+        storeClientCopy(asState->sc, e, asState->storeReadBuffer.legacyReadRequest(result.offset + result.length), asHandleReply, asState);
```

Comment:
The combination of positive leftoversz handled a bit earlier and this `if` statement does not work well -- we cannot do nothing if we have leftovers at EOF. We must process those leftovers instead.

The official code also looks buggy or confusing to me in this area -- how can we even have true leftovers if we only break on seemingly final NUL or buffer end?! (Rhetorical) However, again, I feel that we become partially responsible for this mess by changing copy() sequence termination condition.

Reply:
>  We must process those leftovers instead.

Done.

> we must delete asState when we are done processing available input and we are at EOF.

Done.

> check other recently upgraded special Store clients

I fixed a similar problem in net_db.cc

Reply:
> The official code also looks buggy or confusing to me in this area

Agreed. The official code discarded leftovers from the previous asHandleReply() call when asHandleReply() was called with zero StoreIOBuffer.